### ISSUE TITLE
Multiple daemon instances on docker

### ISSUE THREAD

**Issue by mcadam** [bondagit/aes67-linux-daemon#37]
Hey, I am trying to run multiple instances of the aes67 daemon using docker containers, so I can have multiple inputs on the same machine and avoid having multiple virtual machines.

I am not sure it's actually possible because here we are playing with a kernel module and docker container don't have full access to those but almost.

What I did so far, is build and load the kernel on the host then use the privileged option on the container to have access to the kernel module, this seems to work but then I am getting a connection refused error, apparently coming from the network ?
The container get its own IP and 2 interfaces eth0 and lo so I thought it could have worked but maybe I am missing something.

Here are the logs when starting the container using the ./run_test.sh script, I tried different configurations, using the eth0 interface instead of loopback but same result.

```
❯ docker run -it --privileged --group-add (getent group audio | cut -d: -f3) --cap-add=all -v /lib/modules:/lib/modules aes-daemon
Compiling tools ...
g++    -c -o check.o check.cc
g++   check.o   -o check
g++    -c -o createtest.o createtest.cc
g++   createtest.o   -o createtest
Creating test file ...
test file created
Creating configuration files ..
net.ipv4.igmp_max_memberships = 66
Starting PTP master ...
Starting AES67 daemon ...
Waiting for PTP slave to sync ...
[2021-03-31 13:55:49.100925] [0x00007ff299902600] [debug]   main:: initializing daemon
[2021-03-31 13:55:49.101064] [0x00007ff299902600] [debug]   driver_handler:: sending command code 20 data len 0
[2021-03-31 13:55:49.101254] [0x00007ff299902600] [fatal]   main:: fatal exception error: send_to: Connection refused
[2021-03-31 13:55:49.101268] [0x00007ff299902600] [info]    main:: end
daemon exiting with code: 1
```

Any ideas if its even possible and where would the error connection refused coming from?

Thanks

**Comment by mcadam** [comment]
I forgot to add that if I run the same docker command with the option to use the host network, it works correctly.
So somehow its missing something when using the docker network, not sure what.

Command that works:
```
❯ docker run -it --privileged --network=host aes-daemon
```

**Comment by bondagit** [comment]
Would we able to run multiple instances of the daemon on the same host with multiple network interfaces?
No this is not possible because the daemon uses a single device driver instance to handle the virtual audio part and this can work on a single network interface only.

**Comment by mcadam** [comment]
Thanks for the quick reply. I realised I might not need all that in the end if I can route the channels through different applications using alsa configuration.
I have seen an other closed issue of somebody doing just that.

How many channels total do we have available with the daemon?

**Comment by bondagit** [comment]
> I can route the channels through different applications using alsa configuration.

Yes, this is the way to go. You can use dmix and dsnoop ALSA plug-ins to route the channels to different players and recorders.
The theoretical max number of channels is 64 but depending on your platform you the real limit can be lower than that. Try to use the platform test to verify your real capacity with simultaneous playback and recording.
Otherwise you can activate multiple playbacks in parallel one after the other on your dmix channels. When you see underrun events you know you are approaching the limit.

**Comment by mcadam** [comment]
How do you check if you have underrun events on your host?

**Comment by bondagit** [comment]
> How do you check if you have underrun events on your host?

If you use aplay for example the underrun is printed on the console.

**Comment by bondagit** [comment]
> I suppose a card like this would be much more performant than using the daemon, correct?
> 
> https://www.digigram.com/production-and-sound-cards/lx-dante-ip-audio-network-multichannel-sound-card/

Because of the real time requirement that comes with AES67 I would say that dedicated HW is usually a more reliable solution expecially with a very high number of channels.

**Comment by mcadam** [comment]
Thanks for the replies again, really helpful. 

The machine I am going to use for this would be quite powerful, probably 32 or 64 CPUs with a lot of memory and I would only need maybe 20/30 channels so I reckon the daemon might be ok.
I will run some tests and see how it goes, if its struggling I will go with the dedicated HW.

Do you have some daemon way of running the daemon and the clock, maybe some systemd services configurations? I havent seen something like this in the repo?

**Comment by bondagit** [comment]
No, I don't have anything ready at the moment.

**Comment by BannedPatriot** [comment]
@bondagit where does that 64 channel metric come from? What samplerate does that assume? Does the patched driver remove the 8 channel limit / device check of the 'public build'? If so, are the limits simply system overhead? Can multi-CPU systems with 10Gbps networking exceed 64 channels?

If the check is not removed from the driver, any recomendations on a low-cost device that will unlock it? Thanks.
